pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -356,44 +360,84 @@ func validatePipelineParameterVariables(tasks []PipelineTask, params []ParamSpec
 		}
 	}
 
-	return validatePipelineVariables(tasks, "params", parameterNames, arrayParameterNames)
+	return validatePipelineParametersVariables(tasks, "params", parameterNames, arrayParameterNames)
 }
 
-func validatePipelineVariables(tasks []PipelineTask, prefix string, paramNames sets.String, arrayParamNames sets.String) *apis.FieldError {
+func validatePipelineParametersVariables(tasks []PipelineTask, prefix string, paramNames sets.String, arrayParamNames sets.String) *apis.FieldError {
 	for _, task := range tasks {
-		for _, param := range task.Params {
-			if param.Value.Type == ParamTypeString {
-				if err := validatePipelineVariable(fmt.Sprintf("param[%s]", param.Name), param.Value.StringVal, prefix, paramNames); err != nil {
-					return err
-				}
-				if err := validatePipelineNoArrayReferenced(fmt.Sprintf("param[%s]", param.Name), param.Value.StringVal, prefix, arrayParamNames); err != nil {
+		if err := validateParametersVariablesInParameters(task.Params, prefix, paramNames, arrayParamNames); err != nil {
+			return err
+		}
+		if err := validateParametersVariablesInWhenExpressions(task.WhenExpressions, prefix, paramNames, arrayParamNames); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func validateParametersVariablesInWhenExpressions(whenExpressions WhenExpressions, prefix string, paramNames sets.String, arrayParamNames sets.String) *apis.FieldError {
+	name, path := "task when expression", "pipelinespec.when"
+	for _, we := range whenExpressions {
+		if err := validatePipelineStringVariable(fmt.Sprintf("when.input[%s]", we.Input), we.Input, prefix, name, path, paramNames, arrayParamNames); err != nil {
+			return err
+		}
+		for _, val := range we.Values {
+			if err := validatePipelineStringVariable(fmt.Sprintf("when.values[%s]", val), val, prefix, name, path, paramNames, arrayParamNames); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func validateParametersVariablesInParameters(params []Param, prefix string, paramNames sets.String, arrayParamNames sets.String) *apis.FieldError {
+	name, path := "task parameter", "pipelinespec.params"
+	for _, param := range params {
+		if param.Value.Type == ParamTypeString {
+			if err := validatePipelineStringVariable(fmt.Sprintf("param[%s]", param.Name), param.Value.StringVal, prefix, name, path, paramNames, arrayParamNames); err != nil {
```

Comment:
similar to the when expression, `param` here is not needed, results in `pipelinespec.params.param[b-param]` and on line 402.